plugins/artifacts/artifacts.go

```diff
@@ -1,143 +1,315 @@
-			artifact.Expires = tp.context.TaskInfo.Expires
-		}
-		switch artifact.Type {
-		case "directory":
-			err := result.ExtractFolder(artifact.Path, tp.createUploadHandler(artifact.Name, artifact.Expires))
-			if err != nil {
-				if tp.errorHandled(artifact.Name, artifact.Expires, err) {
-					nonFatalErrs = append(nonFatalErrs, runtime.NewMalformedPayloadError(err.Error()))
-					continue
-				}
-				return false, err
-			}
-		case "file":
-			fileReader, err := result.ExtractFile(artifact.Path)
-			if err != nil {
-				if tp.errorHandled(artifact.Name, artifact.Expires, err) {
-					nonFatalErrs = append(nonFatalErrs, runtime.NewMalformedPayloadError(err.Error()))
-					continue
-				}
-				return false, err
-			}
-			err = tp.attemptUpload(fileReader, artifact.Path, artifact.Name, artifact.Expires)
-			if err != nil {
-				return false, err
-			}
+func (tp *taskPlugin) processFile(result engines.ResultSet, a artifact) {
+	debug("extracting file from path: %s", a.Path)
+	r, err := result.ExtractFile(a.Path)
+	// Always close reader, if one is returned
+	defer func() {
+		if r != nil {
+			r.Close()
 		}
+	}()
+
+	// If feature isn't supported this is malformed-payload
+	if err == engines.ErrFeatureNotSupported {
+		e := runtime.NewMalformedPayloadError(
+			"Artifact extraction is not supported in current configuration of this workerType",
+		)
+		tp.mErrors.Lock()
+		tp.errors = append(tp.errors, e)
+		tp.mErrors.Unlock()
+		return
 	}
 
-	if len(nonFatalErrs) > 0 {
-		// Only report an exception if the command executed successfully.
-		// The logic behind this it is that no artifact upload failure is
-		// expected for a succeeded run, but failed tasks might be missing
-		// some artifacts.
+	// If resource isn't found, task should fail and we print a message to task log
+	if err == engines.ErrResourceNotFound {
+		tp.failed.Set(true)
 		if result.Success() {
-			return false, runtime.MergeMalformedPayload(nonFatalErrs...)
+			// Only complain about missing artifacts, if the task was successful
+			tp.context.LogError(fmt.Sprintf("Artifact '%s' was not found.", a.Path))
 		}
+		tp.context.CreateErrorArtifact(runtime.ErrorArtifact{
+			Name:    a.Name,
+			Reason:  reasonFileMissing,
+			Message: fmt.Sprintf("No file was found at path: '%s' on worker", a.Path),
+			Expires: a.Expires,
+		})
+		return
+	}
+
+	// If malformed-payload error, then path is invalid we wrap and return
+	if e, ok := runtime.IsMalformedPayloadError(err); ok {
+		tp.mErrors.Lock()
+		tp.errors = append(tp.errors, runtime.NewMalformedPayloadError(
+			"Invalid path: '", a.Path, "' reason: ", strings.Join(e.Messages(), ";"),
+		))
+		tp.mErrors.Unlock()
+		return
+	}
 
-		return false, nil
+	// If non-fatal internal error, we want to propagate (what else can we do)
+	if err == runtime.ErrNonFatalInternalError {
+		tp.nonFatalErr.Set(true) // Propagate the non-fatal error
+		tp.monitor.Warn("Received ErrNonFatalInternalError from ResultSet.ExtractFile()")
```

Comment:
it doesn't... `err == runtime.ErrNonFatalInternalError`, ie. it's a specific value.